org.eclipse.xtext/src/org/eclipse/xtext/formatting2/regionaccess/internal/NodeModelBasedRegionAccess.java

```diff
@@ -46,6 +46,9 @@ public String textForOffset(int offset, int length) {
 
 	@Override
 	public AbstractEObjectRegion regionForEObject(EObject obj) {
+		if (obj != null && obj.eResource() != resource) {
+			throw new IllegalArgumentException("ITextRegionAccess.regionForEObject may not be called for resource external EObjects.");
+		}
```

Comment:
for performance reasons, it's better to do the check `obj != null && obj.eResource() != resource` only when `eObjectToTokens.get(obj)` returns null.